sql/analyzer/validation_rules.go

```diff
@@ -512,12 +456,59 @@ func validateIntervalUsage(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Sco
 	})
 
 	if invalid {
-		return nil, transform.SameTree, ErrIntervalInvalidUse.New()
+		return nil, transform.SameTree, analyzererrors.ErrIntervalInvalidUse.New()
 	}
 
 	return n, transform.SameTree, nil
 }
 
+func validateStarExpressions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	var err error
+	transform.Inspect(n, func(n sql.Node) bool {
+		if n == nil {
+			return false
+		}
+
+		if plan.IsDDLNode(n) {
+			return false
+		}
+
+		if er, ok := n.(sql.Expressioner); ok {
+			for _, e := range er.Expressions() {
+				// An expression consisting of just a * is allowed.
+				if _, s := e.(*expression.Star); s {
+					return false
+				}
+				// Otherwise, * can only be used inside acceptable aggregation functions.
+				// Detect any uses of * outside such functions.
+				sql.Inspect(e, func(e sql.Expression) bool {
+					if e == nil {
+						return err == nil
+					}
+					if err != nil {
+						return false
+					}
+					switch e.(type) {
+					case *expression.Star:
+						err = analyzererrors.ErrStarUnsupported.New()
+						return false
+					case *aggregation.Count, *aggregation.CountDistinct, *aggregation.JsonArray:
```

Comment:
if `sql.Inspect` doesn't find the nested star, should probably use `transform.InspectExpr` or one of the others that walks children. something like this is easy if you only have one error type:
```golang
foundStar = transform.InspectUp(n, func(n sql.Node) bool {
    return transform.InspectExpr(e, func(e sql.Expression) bool {
        _, ok := e.(*expression.Star)
        return ok
    })
})
if foundStar {
    return analyzererrors.ErrStarUnsupported.New()
}
```

Reply:
That does look like it's a bit more terse wrt error handling, but ultimately equivalent. I think sql.Inspect works just fine here.

Reply:
Ah ok i misunderstood the original intent. Might be worth adding a docstring with a sentence describing valid vs invalid * uses?

Reply:
Good point. This needs a docstring. I added one.